src/harness/fourslash.ts

```diff
@@ -1790,17 +1790,20 @@ namespace FourSlash {
                 this.editScriptAndUpdateMarkers(fileName, offsetStart, offsetEnd, edit.newText);
                 const editDelta = edit.newText.length - edit.span.length;
                 if (offsetStart <= this.currentCaretPosition) {
-                    this.currentCaretPosition += editDelta;
+                    if (offsetEnd <= this.currentCaretPosition) {
+                        // The entirety of the edit span falls before the caret position, shift the caret accordingly
+                        this.currentCaretPosition += editDelta;
+                    }
+                    else {
+                        // The span being replaced includes the caret position, place the caret at the beginning of the span
+                        this.currentCaretPosition = offsetStart;

```

Comment:
In practice, the test cases only hit this when the caret is already at the start of the span (i.e. ```this.currentCaretPosition === offsetStart``` already). So it makes sense not to move it in that case. When the caret is in the middle, it's kind of arbitrary which way to go. I chose the route with the least logic. (Note: I checked VS it does something different altogether. When the caret is in the middle, it only repositions the caret if it would end up outside the range of the new text. Which makes sense. But we literally never hit this case in tests, so I'm not going to spend any more time on it).